lib/resty/wasmx/shm.lua

```diff
@@ -149,144 +155,163 @@ local _metric_type_set = {
-        page_total = 0,
-        cpage_total = ffi_new("ngx_uint_t [1]"),
-        ckeys = page_size and ffi_new("ngx_str_t *[?]", page_size) or _kbuf
+        clast_index = ffi_new("ngx_uint_t[1]"),
+        ccur_index = ffi_new("ngx_uint_t[1]"),
+        ckeys = page_size and ffi_new("ngx_str_t *[?]", page_size) or _kbuf,
     }
 
     return key_iterator, ctx
 end
 
 
-local function get_keys(zone, max)
-    if max ~= nil and type(max) ~= "number" then
-        error("max must be number", 2)
-    end
+local function shm_get_keys(zone, max_count)
+    if max_count == nil then
+        max_count = DEFAULT_KEYS_PAGE_SIZE
 
-    local shm = zone[WASM_SHM_KEY]
-    local nkeys
-    local keys
-    local start = 0
-    local ctotal = ffi_new("ngx_uint_t [1]")
+    elseif type(max_count) ~= "number" then
+        error("max_count must be a number", 2)
 
-    lock_shm(zone)
+    elseif max_count < 0 then
+        error("max_count must be >= 0", 2)
+    end
 
-    if max == 0 then
-        local rc = C.ngx_wa_ffi_shm_iterate_keys(shm, start, 0, nil, ctotal)
-        nkeys = tonumber(ctotal[0])
+    shm_lock(zone)
 
-        if rc == FFI_DECLINED then
-            unlock_shm(zone)
-            return {}
-        end
+    local shm = zone[WASM_SHM_KEY]
+    local nkeys = tonumber(C.ngx_wa_ffi_shm_kv_nelts(shm))
+    if nkeys == 0 then
+        shm_unlock(zone)
+        return {}
+    end
 
-    else
-        nkeys = max and max or DEFAULT_KEY_BATCH_SIZE
+    if max_count > 0 then
+        nkeys = min(nkeys, max_count)
     end
 
+    local ctotal = ffi_new("ngx_uint_t[1]")
     local ckeys = ffi_new("ngx_str_t *[?]", nkeys)
 
-    C.ngx_wa_ffi_shm_iterate_keys(shm, start, nkeys, ckeys, ctotal)
+    C.ngx_wa_ffi_shm_iterate_keys(shm, nkeys, nil, ctotal, ckeys)
 
-    unlock_shm(zone)
+    shm_unlock(zone)
 
     local total = tonumber(ctotal[0])
-    local keys = table_new(0, total)
+    local keys = new_tab(0, total)
+    assert(total == nkeys)
 
     for i = 1, total do
-        local ckey = ckeys[i - 1]
-        keys[i] = ffi_str(ckey.data, ckey.len)
+        keys[i] = ffi_str(ckeys[i - 1].data, ckeys[i - 1].len)
     end
 
     return keys
 end
 
 
-local function get_kv_value(zone, key)
+local function shm_kv_get(zone, key)
     if type(key) ~= "string" then
-        error("key must be string", 2)
+        error("key must be a string", 2)
     end
 
     local shm = zone[WASM_SHM_KEY]
     local cname = ffi_new("ngx_str_t", { data = key, len = #key })
     local cvalue = ffi_new("ngx_str_t *[1]")
-    local ccas = ffi_new("uint32_t [1]")
+    local ccas = ffi_new("uint32_t[1]")
 
-    local rc = C.ngx_wa_ffi_shm_get_kv_value(shm, cname, cvalue, ccas)
+    local rc = C.ngx_wa_ffi_shm_kv_get(shm, cname, cvalue, ccas)
     if rc == FFI_DECLINED then
-        return nil, nil, "key not found"
+        return nil
+    end
+
+    if rc ~= FFI_OK then
+        return nil, "unknown error"
```

Comment:
I don't understand why you changed the logic around the return codes. 

`ngx_wa_ffi_shm_kv_get` returns whatever `ngx_wa_shm_kv_get_locked` returns. 
`ngx_wa_shm_kv_get_locked` returns either `NGX_DECLINED` when a key is not found or `NGX_OK` when it is found.

Reply:
There's really no space for an unknown error.

Reply:
Throughout our codebase we assert return codes of all underlying C APIs for defensive reasons, future-proofing, etc. The FFI should be no exception. I originally had `assert(rc == FFI_OK)` but switched to a less-harsh standard error. Assertions provide a nice stacktrace to investigate but standard errors may avoid the error-out for non-critical code paths. This really is all about being defensive so I thought a non-intrusive error would be better.

Reply:
Got it. It did occur to me that it could come from defensive programming but it wasn't clear if that was standard we should follow throughout the FFI layer.

Thank you for the explanation!

Reply:
I am thinking of switching these back to assertions, and perhaps only having these assertions enabled in debug mode (which iirc is possible to check somehow as I think I did it previously).